### PR TITLE
8316746: Top of lock-stack does not match the unlocked object

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
It took me quite a long journey to figure out that C2 OSR goes wrong only in the test "vmTestbase/nsk/jdi/StepEvent" because the interpreter fills the slots in a different order as expected by C2. (Interpreter and C1 don't care about it.)
I've reimplemented the search loop in `monitorenter`, improved comments and cleaned up the related code a bit. The test is passing with this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316746](https://bugs.openjdk.org/browse/JDK-8316746): Top of lock-stack does not match the unlocked object (**Bug** - P2)

### Reviewers
 * [Richard Reingruber](https://openjdk.org/census#rrich) (@reinrich - **Reviewer**) ⚠️ Review applies to [5011761f](https://git.openjdk.org/jdk/pull/16406/files/5011761fd5667a4d367037ef43028da82143d4f0)
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**) ⚠️ Review applies to [56209f27](https://git.openjdk.org/jdk/pull/16406/files/56209f27a9986f0ad0db25bc12d4d11434a71881)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16406/head:pull/16406` \
`$ git checkout pull/16406`

Update a local copy of the PR: \
`$ git checkout pull/16406` \
`$ git pull https://git.openjdk.org/jdk.git pull/16406/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16406`

View PR using the GUI difftool: \
`$ git pr show -t 16406`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16406.diff">https://git.openjdk.org/jdk/pull/16406.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16406#issuecomment-1783409041)

### CODE REVIEW

**Comment by dean-long** [comment]
Mystery solved!

**Comment by TheRealMDoerr** [comment]
> Mystery solved!

Yeah! Thanks again for your help! I hope I didn't steal too much of your time and that it was at least interesting for everybody.

**Review by reinrich — COMMENTED** [review]
Hi Martin,

the order of [`BasicObjectLocks` in the interpreter is specified](https://github.com/openjdk/jdk/blob/7a7b1e5a920d71ab717d8993c9258a01f1074a48/src/hotspot/share/runtime/frame.hpp#L368-L375) if the locking operations are well-formed.

the comment on the PPC implementation of [`frame::interpreter_frame_monitor_end`](https://github.com/openjdk/jdk/blob/7a7b1e5a920d71ab717d8993c9258a01f1074a48/src/hotspot/cpu/ppc/frame_ppc.cpp#L457-L458) is wrong. It should be removed.

**Comment by TheRealMDoerr** [comment]
You're right! Removed the comment from all platforms which still had it.

**Comment by reinrich** [comment]
Thanks. I wasn't even aware that other implementations have it as well :)

**Review by reinrich — APPROVED** [review]
The code in this PR is a lot easier to follow than the baseline version and it looks correct to me.
I tried to understand what's actually wrong with the baseline but gave up. Do you know where precisely the bug is?

Anyway, thanks for fixing this!
Richard.

**Comment by TheRealMDoerr** [comment]
Thanks for the review! The old code simply reuses the first empty slot it finds (from top to bottom). That causes the order to be reversed if e.g. 2 slots are free and we lock 2 objects. The new code uses the free slot which is closer to the bottom first.

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp [hunk 1]
```diff
@@ -4159,79 +4159,70 @@ void TemplateTable::athrow() {
...
+  __ cmpld(CCR0, Rcurrent_monitor, Rbot);
+  __ beq(CCR0, Lallocate_new);
 
   // ------------------------------------------------------------------------------
   // Find a free slot in the monitor block.
-  Label Lfound, Lexit, Lallocate_new;
-  ConditionRegister found_free_slot = CCR0,
-                    found_same_obj  = CCR1,
-                    reached_limit   = CCR6;
+  // Note: The order of the monitors is important for C2 OSR which derives the
```
[the remark this answers is not in the document]
Comment (TheRealMDoerr) [reply src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:4191]: Done.

[the remark this answers is not in the document]
Comment (reinrich) [reply src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:4191]: Thanks.

**Comment by TheRealMDoerr** [comment]
@TobiHartmann: You were also interested in a test. I have included one which reproduces the issue on PPC64 without this fix.

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp [hunk 2]
```diff
@@ -4159,79 +4159,70 @@ void TemplateTable::athrow() {
...
   // ------------------------------------------------------------------------------
   // Check if we found a free slot.
-  __ bind(Lexit);
-
-  __ addi(Rcurrent_monitor, Rcurrent_obj_addr, -(frame::interpreter_frame_monitor_size_in_bytes()) - in_bytes(BasicObjectLock::obj_offset()));
-  __ addi(Rcurrent_obj_addr, Rcurrent_obj_addr, - frame::interpreter_frame_monitor_size_in_bytes());
+  __ cmpdi(CCR0, Rfree_slot, 0);
+  __ beq(CCR0, Lallocate_new);
+  __ mr(Rcurrent_monitor, Rfree_slot);
   __ b(Lfound);
```
Comment (RealLucy) [inline src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:4219]: Wouldn't it be nice to use `Rfree_slot` from here on and save the move to `Rcurrent_monitor` and the unconditional branch?
```
  __ cmpdi(CCR0, Rfree_slot, 0);
  __ bne(CCR0, Lfound);

```

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp [hunk 3]
```diff
@@ -4159,79 +4159,70 @@ void TemplateTable::athrow() {
...
 
   // We didn't find a free BasicObjLock => allocate one.
   __ align(32, 12);
   __ bind(Lallocate_new);
   __ add_monitor_to_stack(false, Rscratch1, Rscratch2);
   __ mr(Rcurrent_monitor, R26_monitor);
```
Comment (RealLucy) [inline src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:4225]: If you do the above optimization, this instruction would need to change to 
` __ mr(Rfree_slot, R26_monitor);`

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp [hunk 4]
```diff
@@ -4241,7 +4232,7 @@ void TemplateTable::monitorenter() {
   // The object has already been popped from the stack, so the expression stack looks correct.
   __ addi(R14_bcp, R14_bcp, 1);
 
-  __ std(Robj_to_lock, 0, Rcurrent_obj_addr);
+  __ std(Robj_to_lock, in_bytes(BasicObjectLock::obj_offset()), Rcurrent_monitor);
   __ lock_object(Rcurrent_monitor, Robj_to_lock);
```
Comment (RealLucy) [inline src/hotspot/cpu/ppc/templateTable_ppc_64.cpp:4236]: `Rcurrent_monitor` would need to be replaced by `Rfree_slot` in the above two lines as well.

**Review by RealLucy — APPROVED** [review]
Changes look good to me. 
I added one minor optimization suggestion you may want to consider. It is not relevant for approval.

File: src/hotspot/cpu/ppc/templateTable_ppc_64.cpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/ppc/templateTable_ppc_64.cpp]: Good idea. Changed.

**Comment by TheRealMDoerr** [comment]
Tests are green.
/integrate

**Pull request closed without merging**